<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>TOPMed Analysis Workshop</title>
  <meta name="description" content="Course materials for the TOPMed analysis workshop">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="TOPMed Analysis Workshop" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for the TOPMed analysis workshop" />
  <meta name="github-repo" content="UW-GAC/topmed_workshop_2018" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="TOPMed Analysis Workshop" />
  
  <meta name="twitter:description" content="Course materials for the TOPMed analysis workshop" />
  



<meta name="date" content="2018-05-11">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html">
<link rel="next" href="analysis-pipeline.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>2</b> GDS format</a><ul>
<li class="chapter" data-level="2.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>2.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="2.2" data-path="gds-format.html"><a href="gds-format.html#exercises"><i class="fa fa-check"></i><b>2.2</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>3</b> Computing a GRM</a></li>
<li class="chapter" data-level="4" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>4</b> PC-Relate</a><ul>
<li class="chapter" data-level="4.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>4.1</b> KING</a></li>
<li class="chapter" data-level="4.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>4.2</b> PC-AiR</a></li>
<li class="chapter" data-level="4.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>4.3</b> PC-Relate</a></li>
<li class="chapter" data-level="4.4" data-path="pc-relate.html"><a href="pc-relate.html#comparison-with-pedigree"><i class="fa fa-check"></i><b>4.4</b> Comparison with pedigree</a></li>
<li class="chapter" data-level="4.5" data-path="pc-relate.html"><a href="pc-relate.html#exercise"><i class="fa fa-check"></i><b>4.5</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html"><i class="fa fa-check"></i><b>5</b> Phenotype Harmonization</a><ul>
<li class="chapter" data-level="5.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#import-phentoype-files-into-r"><i class="fa fa-check"></i><b>5.1</b> Import phentoype files into R</a></li>
<li class="chapter" data-level="5.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#run-some-quick-diagnostics-on-the-files"><i class="fa fa-check"></i><b>5.2</b> Run some quick diagnostics on the files</a></li>
<li class="chapter" data-level="5.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#use-using-mixed-models-to-compare-studies"><i class="fa fa-check"></i><b>5.3</b> Use using mixed models to compare studies</a><ul>
<li class="chapter" data-level="5.3.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#create-an-annotated-data-frame"><i class="fa fa-check"></i><b>5.3.1</b> Create an Annotated Data Frame</a></li>
<li class="chapter" data-level="5.3.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#obtain-the-genetic-relatedness-matrix"><i class="fa fa-check"></i><b>5.3.2</b> Obtain the genetic relatedness matrix</a></li>
<li class="chapter" data-level="5.3.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-mixed-model-without-harmonization-unit"><i class="fa fa-check"></i><b>5.3.3</b> Fit a mixed model without harmonization unit</a></li>
<li class="chapter" data-level="5.3.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-model-with-harmonization-unit"><i class="fa fa-check"></i><b>5.3.4</b> Fit a model with harmonization unit</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#final-considerations"><i class="fa fa-check"></i><b>5.4</b> Final considerations</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><i class="fa fa-check"></i><b>6</b> Finding unharmonized TOPMed study phenotypes on dbGaP</a><ul>
<li class="chapter" data-level="6.1" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#ways-to-get-topmed-phenotype-data"><i class="fa fa-check"></i><b>6.1</b> Ways to get TOPMed phenotype data</a></li>
<li class="chapter" data-level="6.2" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#dbgap-accession-lingo"><i class="fa fa-check"></i><b>6.2</b> dbGaP accession lingo</a></li>
<li class="chapter" data-level="6.3" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#dbgap-advanced-search-tools"><i class="fa fa-check"></i><b>6.3</b> dbGaP advanced search tools</a><ul>
<li class="chapter" data-level="6.3.1" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#entrez-advanced-search"><i class="fa fa-check"></i><b>6.3.1</b> <a href="https://www.ncbi.nlm.nih.gov/gap/advanced">Entrez advanced search</a></a></li>
<li class="chapter" data-level="6.3.2" data-path="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html"><a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html#faceted-advanced-search"><i class="fa fa-check"></i><b>6.3.2</b> <a href="https://www.ncbi.nlm.nih.gov/projects/gapsolr/facets.html">Faceted advanced search</a></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>7</b> Association tests</a><ul>
<li class="chapter" data-level="7.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>7.1</b> Null model</a></li>
<li class="chapter" data-level="7.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>7.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="7.3" data-path="association-tests.html"><a href="association-tests.html#exercises-1"><i class="fa fa-check"></i><b>7.3</b> Exercises</a></li>
<li class="chapter" data-level="7.4" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>7.4</b> Sliding window tests</a><ul>
<li class="chapter" data-level="7.4.1" data-path="association-tests.html"><a href="association-tests.html#exercise-1"><i class="fa fa-check"></i><b>7.4.1</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="association-tests.html"><a href="association-tests.html#annotation-based-aggregate-tests"><i class="fa fa-check"></i><b>7.5</b> Annotation-based aggregate tests</a><ul>
<li class="chapter" data-level="7.5.1" data-path="association-tests.html"><a href="association-tests.html#working-with-variant-annotation"><i class="fa fa-check"></i><b>7.5.1</b> Working with variant annotation</a></li>
<li class="chapter" data-level="7.5.2" data-path="association-tests.html"><a href="association-tests.html#defining-gene-ranges-for-aggregation"><i class="fa fa-check"></i><b>7.5.2</b> Defining “gene” ranges for aggregation</a></li>
<li class="chapter" data-level="7.5.3" data-path="association-tests.html"><a href="association-tests.html#aggregating-topmed-variants-into-genic-units"><i class="fa fa-check"></i><b>7.5.3</b> Aggregating TOPMed variants into genic units</a></li>
<li class="chapter" data-level="7.5.4" data-path="association-tests.html"><a href="association-tests.html#aggregate-unit-for-association-testing-exercise"><i class="fa fa-check"></i><b>7.5.4</b> Aggregate unit for association testing exercise</a></li>
<li class="chapter" data-level="7.5.5" data-path="association-tests.html"><a href="association-tests.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>7.5.5</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="7.5.6" data-path="association-tests.html"><a href="association-tests.html#exercise-2"><i class="fa fa-check"></i><b>7.5.6</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html"><i class="fa fa-check"></i><b>8</b> Analysis Pipeline</a><ul>
<li class="chapter" data-level="8.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-a-local-cluster"><i class="fa fa-check"></i><b>8.1</b> Running on a local cluster</a></li>
<li class="chapter" data-level="8.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-aws-batch"><i class="fa fa-check"></i><b>8.2</b> Running on AWS Batch</a><ul>
<li class="chapter" data-level="8.2.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#log-into-aws-docker-image"><i class="fa fa-check"></i><b>8.2.1</b> Log into AWS docker image</a></li>
<li class="chapter" data-level="8.2.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#cd-to-working-directory-and-create-config-file"><i class="fa fa-check"></i><b>8.2.2</b> cd to working directory and create config file</a></li>
<li class="chapter" data-level="8.2.3" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#print-out-aws-commands-if-executing-the-pipeline"><i class="fa fa-check"></i><b>8.2.3</b> Print out AWS commands if executing the pipeline</a></li>
<li class="chapter" data-level="8.2.4" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-pipeline"><i class="fa fa-check"></i><b>8.2.4</b> Execute the pipeline</a></li>
<li class="chapter" data-level="8.2.5" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#monitor-the-jobs"><i class="fa fa-check"></i><b>8.2.5</b> Monitor the jobs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="analysis-commons.html"><a href="analysis-commons.html"><i class="fa fa-check"></i><b>9</b> Analysis Commons</a><ul>
<li class="chapter" data-level="9.1" data-path="analysis-commons.html"><a href="analysis-commons.html#outline"><i class="fa fa-check"></i><b>9.1</b> Outline</a></li>
<li class="chapter" data-level="9.2" data-path="analysis-commons.html"><a href="analysis-commons.html#web-interface-and-running-an-analysis-application"><i class="fa fa-check"></i><b>9.2</b> Web Interface and Running an Analysis Application</a><ul>
<li class="chapter" data-level="9.2.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-1-run-a-single-variant-analysis."><i class="fa fa-check"></i><b>9.2.1</b> Exercise 1) Run a single variant analysis.</a></li>
<li class="chapter" data-level="9.2.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5."><i class="fa fa-check"></i><b>9.2.2</b> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="analysis-commons.html"><a href="analysis-commons.html#command-line-interface"><i class="fa fa-check"></i><b>9.3</b> Command line interface</a><ul>
<li class="chapter" data-level="9.3.1" data-path="analysis-commons.html"><a href="analysis-commons.html#log-in-to-awi"><i class="fa fa-check"></i><b>9.3.1</b> Log in to AWI</a></li>
<li class="chapter" data-level="9.3.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-3-navigate-directories-make-output-directory-examine-files"><i class="fa fa-check"></i><b>9.3.2</b> Exercise 3) Navigate directories, make output directory, examine files</a></li>
<li class="chapter" data-level="9.3.3" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-4-run-single-variant-analysis-from-command-line-using-bash-script"><i class="fa fa-check"></i><b>9.3.3</b> Exercise 4) Run single variant analysis from command line using bash script</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="analysis-commons.html"><a href="analysis-commons.html#writing-your-own-apps"><i class="fa fa-check"></i><b>9.4</b> Writing your own Apps</a><ul>
<li class="chapter" data-level="9.4.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform"><i class="fa fa-check"></i><b>9.4.1</b> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</a></li>
<li class="chapter" data-level="9.4.2" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-6-make-qq-plot"><i class="fa fa-check"></i><b>9.4.2</b> Optional Exercise 6) Make QQ plot</a></li>
<li class="chapter" data-level="9.4.3" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-7-run-conditional-analysis"><i class="fa fa-check"></i><b>9.4.3</b> Optional Exercise 7) Run conditional analysis</a></li>
<li class="chapter" data-level="9.4.4" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set"><i class="fa fa-check"></i><b>9.4.4</b> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">TOPMed Analysis Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="association-tests" class="section level1">
<h1><span class="header-section-number">7</span> Association tests</h1>
<p>Since TOPMed has many studies with related participants, we focus on linear mixed models. Logistic mixed models are also possible using GENESIS, see the <a href="https://www.ncbi.nlm.nih.gov/pubmed/27018471">GMMAT paper</a>.</p>
<div id="null-model" class="section level2">
<h2><span class="header-section-number">7.1</span> Null model</h2>
<p>The first step in an association test is to fit the null model. We will need an <code>AnnotatedDataFrame</code> with phenotypes, and a GRM. We have a sample annotation with a <code>sample.id</code> column matched to the GDS file, and a phenotype file with <code>subject_id</code>. (In this example, we use the 1000 Genomes IDs for both sample and subject ID.) For TOPMed data, it is also important to match by study, as subject IDs are not unique across studies.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># sample annotation</span>
workshop.path &lt;-<span class="st"> &quot;https://github.com/UW-GAC/topmed_workshop_2018/raw/master&quot;</span>
sampfile &lt;-<span class="st"> &quot;sample_annotation.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(sampfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(workshop.path, sampfile), sampfile)
annot &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(sampfile)
<span class="kw">library</span>(Biobase)
<span class="kw">head</span>(<span class="kw">pData</span>(annot))</code></pre></div>
<pre><code>##   sample.id subject.id Population          Population.Description sex
## 1   HG00096    HG00096        GBR British in England and Scotland   M
## 2   HG00097    HG00097        GBR British in England and Scotland   F
## 3   HG00099    HG00099        GBR British in England and Scotland   F
## 4   HG00100    HG00100        GBR British in England and Scotland   F
## 5   HG00101    HG00101        GBR British in England and Scotland   M
## 6   HG00102    HG00102        GBR British in England and Scotland   F
##   status
## 1      0
## 2      1
## 3      0
## 4      1
## 5      0
## 6      0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># phenotypes by subject ID</span>
phenfile &lt;-<span class="st"> &quot;phenotype_annotation.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(phenfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(workshop.path, phenfile), phenfile)
phen &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(phenfile)
<span class="kw">head</span>(<span class="kw">pData</span>(phen))</code></pre></div>
<pre><code>##   subject_id sex age height   study
## 1    HG00096   M  47  165.3 study_1
## 2    HG00102   F  49  169.1 study_1
## 3    HG00112   M  46  167.9 study_1
## 4    HG00114   M  49  169.5 study_1
## 5    HG00115   M  35  161.1 study_1
## 6    HG00116   M  37  182.2 study_1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">varMetadata</span>(phen)</code></pre></div>
<pre><code>##                        labelDescription
## subject_id           subject identifier
## sex                       subject&#39;s sex
## age        age at measurement of height
## height           subject&#39;s height in cm
## study                  study identifier</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># merge sample annotation with phenotypes</span>
<span class="kw">library</span>(dplyr)
dat &lt;-<span class="st"> </span><span class="kw">pData</span>(annot) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">left_join</span>(<span class="kw">pData</span>(phen), <span class="dt">by=</span><span class="kw">c</span>(<span class="st">&quot;subject.id&quot;</span>=<span class="st">&quot;subject_id&quot;</span>, <span class="st">&quot;sex&quot;</span>=<span class="st">&quot;sex&quot;</span>))
meta &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(<span class="kw">varMetadata</span>(annot), <span class="kw">varMetadata</span>(phen)[<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>,,<span class="dt">drop=</span><span class="ot">FALSE</span>])
annot &lt;-<span class="st"> </span><span class="kw">AnnotatedDataFrame</span>(dat, meta)

<span class="co"># load the GRM</span>
data.path &lt;-<span class="st"> &quot;https://github.com/UW-GAC/analysis_pipeline/raw/master/testdata&quot;</span>
grmfile &lt;-<span class="st"> &quot;grm.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(grmfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(data.path, grmfile), grmfile)
grm &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(grmfile)
<span class="co"># the row and column names of the covariance matrix must be set to sample.id</span>
<span class="kw">rownames</span>(grm<span class="op">$</span>grm) &lt;-<span class="st"> </span><span class="kw">colnames</span>(grm<span class="op">$</span>grm) &lt;-<span class="st"> </span>grm<span class="op">$</span>sample.id</code></pre></div>
<p>We will test for an association between genotype and height, adjusting for sex, age, and study as covariates. If the sample set involves multiple distinct groups with different variances for the phenotype, we recommend allowing the model to use heterogeneous variance among groups with the parameter <code>group.var</code>. We saw in a previous exercise that the variance differs by study.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GENESIS)
nullmod &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(annot, <span class="dt">outcome=</span><span class="st">&quot;height&quot;</span>, <span class="dt">covars=</span><span class="kw">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;age&quot;</span>, <span class="st">&quot;study&quot;</span>), 
                        <span class="dt">cov.mat=</span>grm<span class="op">$</span>grm, <span class="dt">group.var=</span><span class="st">&quot;study&quot;</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>We also recommend taking an inverse normal transform of the residuals and refitting the model. This is done separately for each group, and the transformed residuals are rescaled. See the full procedure in the<br />
<a href="https://github.com/UW-GAC/analysis_pipeline#association-testing">pipeline documenation</a>.</p>
</div>
<div id="single-variant-tests" class="section level2">
<h2><span class="header-section-number">7.2</span> Single-variant tests</h2>
<p>Single-variant tests are the same as in GWAS. We use the <code>assocTestSingle</code> function in GENESIS. First, we have to create a <code>SeqVarData</code> object including both the GDS file and the sample annotation containing phenotypes. We then create a <code>SeqVarBlockIterator</code> object to iterate over blocks of variants.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqVarTools)
gdsfile &lt;-<span class="st"> &quot;1KG_phase3_subset_chr1.gds&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(gdsfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(data.path, gdsfile), gdsfile)
gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)
seqData &lt;-<span class="st"> </span><span class="kw">SeqVarData</span>(gds, <span class="dt">sampleData=</span>annot)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod)</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc)</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs         freq       Score
## 1          1   1  970546            1  1126 0.0039964476 -0.11068049
## 2          2   1  985900            1  1126 0.0492895204 -1.23950276
## 3          3   1 1025045            1  1126 0.0004440497 -0.20902154
## 4          4   1 1265550            1  1126 0.0008880995 -0.09909574
## 5          5   1 1472676            1  1126 0.0071047957  0.25276727
## 6          6   1 1735725            1  1126 0.0022202487 -0.10587433
##     Score.SE Score.Stat Score.pval
## 1 0.26797526 -0.4130250 0.67958828
## 2 0.74178030 -1.6709837 0.09472490
## 3 0.09040909 -2.3119527 0.02078029
## 4 0.09771638 -1.0141160 0.31052742
## 5 0.37957709  0.6659181 0.50546342
## 6 0.17644019 -0.6000579 0.54846765</code></pre>
<p>We make a QQ plot to examine the results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
qqPlot &lt;-<span class="st"> </span><span class="cf">function</span>(pval) {
    pval &lt;-<span class="st"> </span>pval[<span class="op">!</span><span class="kw">is.na</span>(pval)]
    n &lt;-<span class="st"> </span><span class="kw">length</span>(pval)
    x &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>n
    dat &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">obs=</span><span class="kw">sort</span>(pval),
                      <span class="dt">exp=</span>x<span class="op">/</span>n,
                      <span class="dt">upper=</span><span class="kw">qbeta</span>(<span class="fl">0.025</span>, x, <span class="kw">rev</span>(x)),
                      <span class="dt">lower=</span><span class="kw">qbeta</span>(<span class="fl">0.975</span>, x, <span class="kw">rev</span>(x)))
    
    <span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(obs))) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(upper)), <span class="dt">color=</span><span class="st">&quot;gray&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(lower)), <span class="dt">color=</span><span class="st">&quot;gray&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_abline</span>(<span class="dt">intercept=</span><span class="dv">0</span>, <span class="dt">slope=</span><span class="dv">1</span>, <span class="dt">color=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">xlab</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="op">-</span>log[<span class="dv">10</span>], <span class="st">&quot;(expected P)&quot;</span>))) <span class="op">+</span>
<span class="st">        </span><span class="kw">ylab</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="op">-</span>log[<span class="dv">10</span>], <span class="st">&quot;(observed P)&quot;</span>))) <span class="op">+</span>
<span class="st">        </span><span class="kw">theme_bw</span>()
}    

<span class="kw">qqPlot</span>(assoc<span class="op">$</span>Score.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_single_qq-1.png" width="672" /></p>
</div>
<div id="exercises-1" class="section level2">
<h2><span class="header-section-number">7.3</span> Exercises</h2>
<ol style="list-style-type: decimal">
<li>Logistic regression: <code>fitNullModel</code> can use a binary phenotype as the outcome variable by specifying the argument <code>family=binomial</code>. Use the <code>status</code> column in the sample annotation to fit a null model for simulated case/control status, with <code>sex</code> and <code>Population</code> as covariates. Refer to the documentation for <code>fitNulModel</code> to see what other parameters need to be changed for a binary outcome. Then run a single-variant test using this model.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nullmod.status &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(annot, <span class="dt">outcome=</span><span class="st">&quot;status&quot;</span>, <span class="dt">covars=</span><span class="kw">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;Population&quot;</span>), 
                               <span class="dt">cov.mat=</span>grm<span class="op">$</span>grm, <span class="dt">family=</span>binomial, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
<span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod.status, <span class="dt">test=</span><span class="st">&quot;Score&quot;</span>)</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc)</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs         freq       Score
## 1          1   1  970546            1  1126 0.0039964476  0.20297260
## 2          2   1  985900            1  1126 0.0492895204 -2.63528215
## 3          3   1 1025045            1  1126 0.0004440497 -0.09788685
## 4          4   1 1265550            1  1126 0.0008880995  0.81055788
## 5          5   1 1472676            1  1126 0.0071047957  0.63907684
## 6          6   1 1735725            1  1126 0.0022202487 -0.45986600
##    Score.SE Score.Stat Score.pval
## 1 0.8333773  0.2435543 0.80757604
## 2 2.6406893 -0.9979524 0.31830245
## 3 0.2944788 -0.3324071 0.73958188
## 4 0.4079077  1.9871110 0.04691009
## 5 1.0775580  0.5930788 0.55312840
## 6 0.6352752 -0.7238847 0.46913653</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Inverse normal transform: use the function <code>nullModelInvNorm</code> to perform an inverse normal transform on the <code>height</code> variable. For each study separately, compute a null model and do the inverse normal transform using just the values for that study. Compare these residuals with the initial residuals you obtained for that study by transforming all studies together.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nullmod.norm.all &lt;-<span class="st"> </span><span class="kw">nullModelInvNorm</span>(nullmod, <span class="dt">cov.mat=</span>grm<span class="op">$</span>grm, <span class="dt">norm.option=</span><span class="st">&quot;all&quot;</span>)</code></pre></div>
<pre><code>## [1]  5.426533e+01  4.280310e+01  1.165095e+02  9.808945e+01 -1.603237e+03
## [6]  7.550181e-03
## [1]  3.410792e+01  3.584897e+01  8.954492e+01  7.624706e+01 -1.603138e+03
## [6]  1.035647e-02
## [1]  2.396106e+01  2.656835e+01  4.184204e+01  3.859662e+01 -1.589464e+03
## [6]  1.687874e-02
## [1]    13.2491648    16.9373053    20.2505893    18.9509796 -1586.8577254
## [6]     0.0311776
## [1]  6.261202e-02  1.472786e+00  4.156538e-01  4.398434e-01 -1.667973e+03
## [6]  1.553884e+00
## [1]     0.1502457     0.4474393     0.5757750     0.5944981 -1591.7256219
## [6]     1.4305344
## [1]     0.2764169     0.5476046     0.6540029     0.6438323 -1587.8624868
## [6]     1.1050621
## [1]     0.3788940     0.5650763     0.6354999     0.6027325 -1586.7797445
## [6]     1.0118797
## [1]     0.4009141     0.5618423     0.6268258     0.5894059 -1586.7486905
## [6]     1.0002449
## [1]     0.4014081     0.5613116     0.6270659     0.5891437 -1586.7486447
## [6]     1.0000002
## [1]     0.4014424     0.5612675     0.6270502     0.5891168 -1586.7486446
## [6]     1.0000000
## [1]     0.4014438     0.5612652     0.6270502     0.5891156 -1586.7486446
## [6]     1.0000000
## [1]     0.4014439     0.5612651     0.6270501     0.5891155 -1586.7486446
## [6]     1.0000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat.all &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample.id=</span>nullmod.norm.all<span class="op">$</span>sample.id, 
                      <span class="dt">resid.norm=</span>nullmod.norm.all<span class="op">$</span>resid.marginal,
                      <span class="dt">study=</span>annot<span class="op">$</span>study,
                      <span class="dt">run=</span><span class="st">&quot;combined&quot;</span>)

nullmod.norm.group &lt;-<span class="st"> </span><span class="kw">nullModelInvNorm</span>(nullmod, <span class="dt">cov.mat=</span>grm<span class="op">$</span>grm, <span class="dt">norm.option=</span><span class="st">&quot;by.group&quot;</span>)</code></pre></div>
<pre><code>## [1]  5.426533e+01  4.280310e+01  1.165095e+02  9.808945e+01 -1.603282e+03
## [6]  7.550782e-03
## [1]  3.410783e+01  3.584937e+01  8.954476e+01  7.624651e+01 -1.603095e+03
## [6]  1.035568e-02
## [1]  2.381328e+01  2.675677e+01  4.188200e+01  3.854148e+01 -1.589110e+03
## [6]  1.688561e-02
## [1]  1.314292e+01  1.711312e+01  2.027537e+01  1.891724e+01 -1.586415e+03
## [6]  3.118595e-02
## [1]  8.468674e-02  1.482857e+00  3.925956e-01  4.194635e-01 -1.668069e+03
## [6]  1.544911e+00
## [1]     0.1798233     0.3693854     0.5469337     0.5675795 -1592.4318992
## [6]     1.4789592
## [1]     0.2965849     0.4915445     0.6366946     0.6250394 -1587.7319467
## [6]     1.1236269
## [1]     0.3823806     0.5525549     0.6327231     0.5971163 -1586.3518912
## [6]     1.0156283
## [1]     0.3993051     0.5657771     0.6265088     0.5870741 -1586.3023605
## [6]     1.0004057
## [1]     0.3991862     0.5666070     0.6268995     0.5872827 -1586.3023287
## [6]     1.0000003
## [1]     0.3991706     0.5666290     0.6268996     0.5873005 -1586.3023286
## [6]     1.0000000
## [1]     0.3991697     0.5666302     0.6268999     0.5873013 -1586.3023286
## [6]     1.0000000
## [1]     0.3991697     0.5666303     0.6268999     0.5873014 -1586.3023286
## [6]     1.0000000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat.group &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample.id=</span>nullmod.norm.group<span class="op">$</span>sample.id, 
                        <span class="dt">resid.norm=</span>nullmod.norm.group<span class="op">$</span>resid.marginal,
                        <span class="dt">study=</span>annot<span class="op">$</span>study,
                        <span class="dt">run=</span><span class="st">&quot;separate&quot;</span>)

dat &lt;-<span class="st"> </span><span class="kw">rbind</span>(dat.all, dat.group)

<span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(study, resid.norm, <span class="dt">fill=</span>run)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>()</code></pre></div>
<p><img src="_main_files/figure-html/exercise_invnorm-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">group_by</span>(study, run) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">summarise</span>(<span class="dt">mean=</span><span class="kw">mean</span>(resid.norm), <span class="dt">var=</span><span class="kw">var</span>(resid.norm))</code></pre></div>
<pre><code>## # A tibble: 6 x 4
## # Groups:   study [?]
##   study   run           mean   var
##   &lt;fct&gt;   &lt;fct&gt;        &lt;dbl&gt; &lt;dbl&gt;
## 1 study_1 combined  0.000354 1.00 
## 2 study_1 separate  0.00121  0.999
## 3 study_2 combined  0.000932 1.00 
## 4 study_2 separate  0.000816 0.999
## 5 study_3 combined -0.00129  1.000
## 6 study_3 separate -0.00203  0.999</code></pre>
</div>
<div id="sliding-window-tests" class="section level2">
<h2><span class="header-section-number">7.4</span> Sliding window tests</h2>
<p>For rare variants, we can do burden tests or SKAT using the GENESIS function <code>assocTestSeqAggregate</code>. We restrict the test to variants with alternate allele frequency &lt; 0.1. (For real data, this threshold would be lower.) We use a flat weighting scheme. We define a sliding window across the genome using a <code>SeqVarWindowIterator</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarWindowIterator</span>(seqData, <span class="dt">windowSize=</span><span class="dv">5000</span>, <span class="dt">windowShift=</span><span class="dv">2000</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;Burden&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(assoc)</code></pre></div>
<pre><code>## [1] &quot;results&quot;     &quot;variantInfo&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr   start     end n.site n.alt n.sample.alt       Score   Score.SE
## 1   1  966001  971000      1     9            9 -0.11068049 0.26797526
## 2   1  982001  987000      1   111          107 -1.23950276 0.74178030
## 3   1 1022001 1027000      1     1            1 -0.20902154 0.09040909
## 4   1 1262001 1267000      1     2            2 -0.09909574 0.09771638
## 5   1 1468001 1473000      1    16           16  0.25276727 0.37957709
## 6   1 1732001 1737000      1     5            5 -0.10587433 0.17644019
##   Score.Stat Score.pval
## 1 -0.4130250 0.67958828
## 2 -1.6709837 0.09472490
## 3 -2.3119527 0.02078029
## 4 -1.0141160 0.31052742
## 5  0.6659181 0.50546342
## 6 -0.6000579 0.54846765</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo)</code></pre></div>
<pre><code>## [[1]]
##   variant.id chr    pos allele.index n.obs        freq weight
## 1          1   1 970546            1  1126 0.003996448      1
## 
## [[2]]
##   variant.id chr    pos allele.index n.obs       freq weight
## 1          2   1 985900            1  1126 0.04928952      1
## 
## [[3]]
##   variant.id chr     pos allele.index n.obs         freq weight
## 1          3   1 1025045            1  1126 0.0004440497      1
## 
## [[4]]
##   variant.id chr     pos allele.index n.obs         freq weight
## 1          4   1 1265550            1  1126 0.0008880995      1
## 
## [[5]]
##   variant.id chr     pos allele.index n.obs        freq weight
## 1          5   1 1472676            1  1126 0.007104796      1
## 
## [[6]]
##   variant.id chr     pos allele.index n.obs        freq weight
## 1          6   1 1735725            1  1126 0.002220249      1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>Score.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_window_burden-1.png" width="672" /></p>
<p>For SKAT, we use the Wu weights.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarWindowIterator</span>(seqData, <span class="dt">windowSize=</span><span class="dv">5000</span>, <span class="dt">windowShift=</span><span class="dv">2000</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;SKAT&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">25</span>))</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr   start     end n.site n.alt n.sample.alt       Q_0     pval_0 err_0
## 1   1  966001  971000      1     9            9  6.317493 0.67958828     0
## 2   1  982001  987000      1   111          107 84.857944 0.09472490     0
## 3   1 1022001 1027000      1     1            1 26.730269 0.02078029     0
## 4   1 1262001 1267000      1     2            2  5.881232 0.31052742     0
## 5   1 1468001 1473000      1    16           16 28.358648 0.50546342     0
## 6   1 1732001 1737000      1     5            5  6.296893 0.54846765     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo)</code></pre></div>
<pre><code>## [[1]]
##   variant.id chr    pos allele.index n.obs        freq   weight
## 1          1   1 970546            1  1126 0.003996448 22.70917
## 
## [[2]]
##   variant.id chr    pos allele.index n.obs       freq   weight
## 1          2   1 985900            1  1126 0.04928952 7.431881
## 
## [[3]]
##   variant.id chr     pos allele.index n.obs         freq   weight
## 1          3   1 1025045            1  1126 0.0004440497 24.73493
## 
## [[4]]
##   variant.id chr     pos allele.index n.obs         freq   weight
## 1          4   1 1265550            1  1126 0.0008880995 24.47255
## 
## [[5]]
##   variant.id chr     pos allele.index n.obs        freq   weight
## 1          5   1 1472676            1  1126 0.007104796 21.06793
## 
## [[6]]
##   variant.id chr     pos allele.index n.obs        freq   weight
## 1          6   1 1735725            1  1126 0.002220249 23.70132</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>pval_<span class="dv">0</span>)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_window_skat-1.png" width="672" /></p>
<div id="exercise-1" class="section level3">
<h3><span class="header-section-number">7.4.1</span> Exercise</h3>
<p>Repeat the previous exercise on logistic regression, this time running a sliding-window test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nullmod.status &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(annot, <span class="dt">outcome=</span><span class="st">&quot;status&quot;</span>, <span class="dt">covars=</span><span class="kw">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;Population&quot;</span>), 
                               <span class="dt">cov.mat=</span>grm<span class="op">$</span>grm, <span class="dt">family=</span>binomial, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
<span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarWindowIterator</span>(seqData, <span class="dt">windowSize=</span><span class="dv">5000</span>, <span class="dt">windowShift=</span><span class="dv">2000</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;SKAT&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">25</span>))</code></pre></div>
<pre><code>## # of selected samples: 1,126</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr   start     end n.site n.alt n.sample.alt       Q_0     pval_0 err_0
## 1   1  966001  971000      1     9            9  6.317493 0.67958828     0
## 2   1  982001  987000      1   111          107 84.857944 0.09472490     0
## 3   1 1022001 1027000      1     1            1 26.730269 0.02078029     0
## 4   1 1262001 1267000      1     2            2  5.881232 0.31052742     0
## 5   1 1468001 1473000      1    16           16 28.358648 0.50546342     0
## 6   1 1732001 1737000      1     5            5  6.296893 0.54846765     0</code></pre>
</div>
</div>
<div id="annotation-based-aggregate-tests" class="section level2">
<h2><span class="header-section-number">7.5</span> Annotation-based aggregate tests</h2>
<p><strong>Note:</strong> the code and libraries in this section are under active development, and are not production-level. It is provided to give workshop participants an example of some of the kinds of analysis tasks that might be performed with TOPMed annotation data. Use the code at your own risk, and be warned that it may break in unexpected ways. Github issues and contributions are welcome!</p>
<p>Analysts generally aggregate rare variants for association testing to decrease multiple testing burden and increase statistical power. They can group variants that fall within arbitrary ranges (such as sliding-windows), or they can group variants with intent. For example, an analyst could aggregate variants that that fall between transcription start sites and stop sites, within coding regions, within regulatory regions, or other genomic features selected from sources like published gene models or position- or transcript-based variant annotation. An analyst could also choose to filter the variants prior or subsequent to aggregation using annotation-based criteria such as functional impact or quality scores.</p>
<p>To demonstrate, we will aggregate a subset of TOPMed variants from chromosome 22. The subset is a portion of TOPMed SNP and indel variants that are also in the 1000 Genomes Project. We will parse an example variant annotation file to select fields of interest, parse a GENCODE .gtf file to define our genic units, and then aggregate the selected variants into the defined genic units.</p>
<div id="working-with-variant-annotation" class="section level3">
<h3><span class="header-section-number">7.5.1</span> Working with variant annotation</h3>
<p>Variants called from the TOPMed data set are annotated using the <a href="https://sites.google.com/site/jpopgen/wgsa">Whole Genome Sequence Annotator (WGSA)</a>. WGSA output files include 359 annotation fields, some of which are themselves lists of annotation values. Thus, individual variants may be annotated with more than 1000 individual fields. WGSA produces different output files for different featues. TOPMed variant annotation includes separate files for SNPs and for indels. The subsetted variant annotation files we will use for this example are available via github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ben.workshop.path &lt;-<span class="st"> &quot;https://github.com/bheavner/topmed_workshop_2017_bh/raw/master&quot;</span>
snpfile &lt;-<span class="st"> &quot;snp.tsv.gz&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(snpfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, snpfile), snpfile)

indelfile &lt;-<span class="st"> &quot;indel.tsv.gz&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(indelfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, indelfile), indelfile)</code></pre></div>
<p>The WGSA output files are tab-separated text files, with one line per annotated variant. Since there are many annotation fields, these files can be unwieldy to work with directly. As an example, the first two lines of the SNP variant annotation file can be previewed within R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">readLines</span>(<span class="st">&quot;snp.tsv.gz&quot;</span>, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<p>The DCC has begun an R package, <em>wgsaparsr</em>, to begin working with WGSA output files. This package is under development, and is available on github at <a href="https://github.com/UW-GAC/wgsaparsr" class="uri">https://github.com/UW-GAC/wgsaparsr</a>. For now, the package can be installed from github using the <em>devtools</em> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(devtools)</span>
<span class="co">#devtools::install_github(&quot;UW-GAC/wgsaparsr@1.0.0.9003&quot;)</span>
<span class="kw">library</span>(wgsaparsr)
<span class="co">#library(tidyverse) # just in case it&#39;s not loaded yet</span>
<span class="kw">library</span>(tibble)
<span class="kw">library</span>(dplyr)
<span class="kw">library</span>(tidyr)
<span class="kw">library</span>(readr)</code></pre></div>
<p><em>wgsaparsr</em> includes a <em>get_fields()</em> function to list the annotation fields available in a WGSA output file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># list all fields in an annotation file: </span>
<span class="kw">get_fields</span>(<span class="st">&quot;snp.tsv.gz&quot;</span>)</code></pre></div>
<p>Only a subset of these annotations may be necessary for a particular association test, and it is unweildy to work with all of them, so it is useful to process the WGSA output file to select fields of interest. The <em>wgsaparsr</em> function <em>parse_to_file()</em> allows field selection by name.</p>
<p>An additional complication in working with the WGSA output files is that some of the annotation fields are transcript-based, rather than position-based. Thus, if a variant locus is within multiple transcripts, those fields will have multiple entries (often separated by a | character). For example, annotation fields such as <code>VEP_ensembl_Transcript_ID</code> may have many values within a single tab-separated field.</p>
<p><em>wgsaparsr::parse_to_file()</em> addresses this by splitting such list-fields into multiple rows. Other annotation fields for that variant are duplicated, and associated columns are filled with the same value for each transcript that a particular variant falls within. A consequence of this approach is that the processed annotation file has more lines than the WGSA output file. In freeze 4, processing expanded the annotation by a factor of about 5 - the 220 million annotations result in a 1-billion row database for subsequent aggregation.</p>
<p><em>wgsaparsr::parse_to_file()</em> reads a snp annotation file, selects specified fields, and expands user-defined transcript-level annotation fields. It produces a tab-separated output file for subsequent analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">desired_columns &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">c</span>(
    <span class="st">&quot;`#chr`&quot;</span>, <span class="co">#NOTE: backtics on #chr because it starts with special character!</span>
    <span class="st">&quot;pos&quot;</span>,
    <span class="st">&quot;ref&quot;</span>,
    <span class="st">&quot;alt&quot;</span>,
    <span class="st">&quot;rs_dbSNP147&quot;</span>,
    <span class="co"># &quot;CADDphred&quot;,</span>
    <span class="st">&quot;CADD_phred&quot;</span>, <span class="co">#NOTE: different than the indel annotation file.</span>
    <span class="st">&quot;VEP_ensembl_Transcript_ID&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Gene_Name&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Gene_ID&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Consequence&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Amino_Acid_Change&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_filter&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_flags&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_info&quot;</span>
    <span class="co"># &quot;1000Gp3_AF&quot; #skipped for the workshop because code doesn&#39;t work with this variable name</span>
    )
    
to_split &lt;-
<span class="st">    </span><span class="kw">c</span>(
    <span class="st">&quot;VEP_ensembl_Consequence&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Transcript_ID&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Gene_Name&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Gene_ID&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Amino_Acid_Change&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_filter&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_flags&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_info&quot;</span>
    )
  
<span class="kw">parse_to_file</span>(<span class="st">&quot;snp.tsv.gz&quot;</span>, <span class="st">&quot;parsed_snp.tsv&quot;</span>, desired_columns, to_split, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Although the output file has fewer columns than the the raw WGSA output file, this .tsv file is not particularly nice to work with in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">readLines</span>(<span class="st">&quot;parsed_snp.tsv&quot;</span>, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<p>However, <em>get_fields()</em> does work on the parsed file to view available fields:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># list all fields in an annotation file: </span>
<span class="kw">get_fields</span>(<span class="st">&quot;parsed_snp.tsv&quot;</span>)</code></pre></div>
<p>The WGSA output files for indel variants differs from the output for SNPs. Some of the field names differ slightly (e.g. “CADDphred” instead of “CADD_phred”), and there are some fields of interest that include feature counts in brackets (e.g. ENCODE_Dnase_cells includes fields like 125{23}). Thus, (for now) <em>wgsaparsr</em> includes <em>parse_indel_to_file()</em>. <em>parse_indel_to_file()</em> is very similar to <em>parse_to_file()</em>, and will likely be incorporated to that function in the near future. The syntax for <em>parse_indel_to_file()</em> is the same as <em>parse_to_file()</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">desired_columns_indel &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">c</span>(
    <span class="st">&quot;`#chr`&quot;</span>, <span class="co">#NOTE: backtics on #chr because it starts with special character!</span>
    <span class="st">&quot;pos&quot;</span>,
    <span class="st">&quot;ref&quot;</span>,
    <span class="st">&quot;alt&quot;</span>,
    <span class="st">&quot;rs_dbSNP147&quot;</span>,
    <span class="st">&quot;CADDphred&quot;</span>,
    <span class="co">#  &quot;CADD_phred&quot;, #NOTE: different than the general annotation file.</span>
    <span class="st">&quot;VEP_ensembl_Transcript_ID&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Gene_Name&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Gene_ID&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Consequence&quot;</span>,
    <span class="st">&quot;VEP_ensembl_Amino_Acid_Change&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_filter&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_flags&quot;</span>,
    <span class="st">&quot;VEP_ensembl_LoF_info&quot;</span>
    <span class="co"># &quot;1000Gp3_AF&quot;#skipped for the workshop because code doesn&#39;t work with this variable name</span>
    )
  
<span class="kw">parse_indel_to_file</span>(<span class="st">&quot;indel.tsv.gz&quot;</span>, <span class="st">&quot;parsed_indel.tsv&quot;</span>, desired_columns_indel, to_split, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Inspection shows that the output format is the same for this function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">readLines</span>(<span class="st">&quot;parsed_indel.tsv&quot;</span>, <span class="dt">n=</span><span class="dv">2</span>)</code></pre></div>
<p>Or as a list,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># list all fields in an annotation file: </span>
<span class="kw">get_fields</span>(<span class="st">&quot;parsed_indel.tsv&quot;</span>)</code></pre></div>
<p>If an analyst wished to filter the list of variants prior to aggregation, the processing code could be modified to apply filters during parsing, or the annotation file could be reprocessed to apply filters at this point. Alternatively, filters can also be applied subsequent to aggregation.</p>
<p>As insurance for this exercise, the parsed files are also available on github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ben.workshop.path &lt;-<span class="st"> &quot;https://github.com/bheavner/topmed_workshop_2017_bh/raw/master&quot;</span>
parsedsnpfile &lt;-<span class="st"> &quot;parsed_snp.tsv&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(parsedsnpfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, parsedsnpfile), parsedsnpfile)

parsedindelfile &lt;-<span class="st"> &quot;parsed_indel.tsv&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(parsedindelfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, parsedindelfile), parsedindelfile)</code></pre></div>
</div>
<div id="defining-gene-ranges-for-aggregation" class="section level3">
<h3><span class="header-section-number">7.5.2</span> Defining “gene” ranges for aggregation</h3>
<p>Aggregation requires definition of the desired aggregation units. As a relatively simple example, we will build a list of genomic ranges corresponding to genes as defined by the <a href="https://www.gencodegenes.org/about.html">GENCODE Project</a>.</p>
<p>The GENCODE Project’s Genomic ENCylopedia Of DNA Elements is available in the well-documented <a href="https://www.gencodegenes.org/data_format.html">.gtf file format</a>. Generally, .gtf files consist of 9 tab-separated fields, some of which may consist of various numbers of key:value pairs.</p>
<p>The DCC has begun an R package, <em>genetable</em>, to parse and work with .gtf files. This package is under development, and is available on github at <a href="https://github.com/UW-GAC/genetable" class="uri">https://github.com/UW-GAC/genetable</a>. For now, the package can be installed from github using the <em>devtools</em> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#library(devtools)</span>
<span class="co">#devtools::install_github(&quot;UW-GAC/genetable&quot;)</span>

<span class="kw">library</span>(genetable)</code></pre></div>
<p>I’ll be working with the gencode release 19 because it’s the last one on GRCh37. The <a href="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz">file</a> can be downloaded via <a href="https://www.gencodegenes.org/releases/19.html" class="uri">https://www.gencodegenes.org/releases/19.html</a>.</p>
<p>In this case, I’ve trimmed the gencode file to include only chromosome 22 feature definitions (since that’s the variant annotation set I’m using for the demo).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtffile &lt;-<span class="st"> &quot;chr22.gtf.gz&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(gtffile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, gtffile), gtffile)

gtf_source &lt;-<span class="st"> &quot;chr22.gtf.gz&quot;</span></code></pre></div>
<p>The details of the <em>genetable</em> package are of less interest for this workshop, so we’ll just use it - we can import and tidy the .gtf file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># import the gtf file to a tidy data frame (a tibble)</span>
gtf &lt;-<span class="st"> </span><span class="kw">import_gencode</span>(gtf_source)

<span class="co"># look at the tibble</span>
<span class="kw">glimpse</span>(gtf)</code></pre></div>
<p>We can see that genomic features are tagged by feature type:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># summarize the number of features by tag.</span>
<span class="kw">summarize_tag</span>(gtf, <span class="dt">tag =</span> <span class="st">&quot;basic&quot;</span>)</code></pre></div>
<p>And we can use these feature type tags to filter the .gtf annotation to extract the starting and ending genomic positions for features of interest, such as features tagged “gene”:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter gtf file to return transcript features tagged basic</span>
basic_transcripts &lt;-<span class="st"> </span><span class="kw">filter_gencode</span>(gtf, <span class="dt">featurearg =</span> <span class="st">&quot;transcript&quot;</span>, <span class="dt">tagarg =</span> <span class="st">&quot;basic&quot;</span>)

<span class="co"># or filter for features == &quot;gene&quot;</span>
genes &lt;-<span class="st"> </span><span class="kw">filter_gencode</span>(gtf, <span class="dt">featurearg =</span> <span class="st">&quot;gene&quot;</span>)

<span class="co"># define the boundaries of the feature of interest</span>
<span class="co"># this can be slow for complicated features</span>
<span class="co">#gene_bounds &lt;- define_boundaries(basic_transcripts, &quot;gene_id&quot;)</span>
gene_bounds &lt;-<span class="st"> </span><span class="kw">define_boundaries</span>(genes, <span class="st">&quot;gene_id&quot;</span>)

<span class="co"># can check the resulting tibble for sanity</span>
<span class="kw">glimpse</span>(gene_bounds)</code></pre></div>
<p>Finally, <em>genetable</em> includes a function to save a self-documented tab separated file containing the filtered .gtf results:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># save to file</span>
note &lt;-<span class="st"> &#39;This file includes starting and ending ranges for feature = &quot;gene&quot; in the gtf file.&#39;</span>
<span class="kw">save_to_file</span>(gene_bounds, <span class="dt">notes =</span> note) <span class="co"># will automatically make file called feature_bounds_DATE.tsv</span></code></pre></div>
<p>As insurance for this exercise, the genic range definitions that I made last week are also available on github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ben.workshop.path &lt;-<span class="st"> &quot;https://github.com/bheavner/topmed_workshop_2017_bh/raw/master&quot;</span>
boundsfile &lt;-<span class="st"> &quot;feature_bounds_20170804.tsv&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(boundsfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, boundsfile), boundsfile)</code></pre></div>
</div>
<div id="aggregating-topmed-variants-into-genic-units" class="section level3">
<h3><span class="header-section-number">7.5.3</span> Aggregating TOPMed variants into genic units</h3>
<p>Now we’ve generated a set of variants with a manaageable number of annotation fields, and defined the desired aggregation units as sets of genomic ranges. The set of variants may be filtered using the annotation fields we’ve chosen (our list is unfiltered in this example).</p>
<p>We’re ready to aggregate the variants by genic units. As insurance, we can start with the same set of inputs by downloading what I generated last week:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ben.workshop.path &lt;-<span class="st"> &quot;https://github.com/bheavner/topmed_workshop_2017_bh/raw/master&quot;</span>

parsed_snp_file &lt;-<span class="st"> &quot;parsed_snp.tsv&quot;</span>
parsed_indel_file &lt;-<span class="st"> &quot;parsed_indel.tsv&quot;</span>
unit_defs_file &lt;-<span class="st"> &quot;feature_bounds_20170804.tsv&quot;</span>

<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(parsed_snp_file)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, parsed_snp_file), parsed_snp_file)

<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(parsed_indel_file)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, parsed_indel_file), parsed_indel_file)

<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(unit_defs_file)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(ben.workshop.path, unit_defs_file), unit_defs_file)</code></pre></div>
<p>Load the tab-separated files to tibbles (data frames) to work with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snps &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(parsed_snp_file, <span class="dt">comment =</span> <span class="st">&quot;#&quot;</span>)
indels &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(parsed_indel_file, <span class="dt">comment =</span> <span class="st">&quot;#&quot;</span>)
unit_defs &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(unit_defs_file, <span class="dt">comment =</span> <span class="st">&quot;#&quot;</span>, <span class="dt">skip =</span> <span class="dv">1</span>)
unit_defs &lt;-<span class="st"> </span><span class="kw">select</span>(unit_defs, <span class="kw">c</span>(gene_id, agg_start, agg_end))</code></pre></div>
<p>There’s probably a nice, fast, vectorized to accomplish this task, but for demonstration purposes, we’ll just loop over the unit_defs and select indels and snps within the genomic ranges of interest:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># make an empty tibble</span>
foo &lt;-<span class="st"> </span><span class="kw">tibble</span>(<span class="dt">group_id=</span><span class="st">&quot;&quot;</span>, <span class="dt">chromosome=</span><span class="st">&quot;&quot;</span>, <span class="dt">position=</span><span class="st">&quot;&quot;</span>, <span class="dt">ref=</span><span class="st">&quot;&quot;</span>, <span class="dt">alt=</span><span class="st">&quot;&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">length</span>(group_id)<span class="op">&gt;</span><span class="dv">1</span>)

<span class="co"># loop over unit defs</span>
<span class="cf">for</span> (rowIndex <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(unit_defs)) {

  <span class="co"># select snps and insert to foo ## SNPs could be filtered here</span>
  snpsToAdd &lt;-<span class="st"> </span><span class="kw">select</span>(snps, <span class="kw">c</span>(chr, pos, ref, alt)) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="kw">between</span>(pos, unit_defs[rowIndex,]<span class="op">$</span>agg_start, unit_defs[rowIndex,]<span class="op">$</span>agg_end)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># This is the line to vectorize</span>
<span class="st">    </span><span class="kw">distinct</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">group_id =</span> unit_defs[rowIndex,]<span class="op">$</span>gene_id)
  
  <span class="cf">if</span> (<span class="kw">nrow</span>(snpsToAdd) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
    foo &lt;-<span class="st"> </span><span class="kw">add_row</span>(
      foo,
      <span class="dt">group_id =</span> snpsToAdd<span class="op">$</span>group_id,
      <span class="dt">chromosome =</span> snpsToAdd<span class="op">$</span>chr,
      <span class="dt">position =</span> snpsToAdd<span class="op">$</span>pos,
      <span class="dt">ref =</span> snpsToAdd<span class="op">$</span>ref,
      <span class="dt">alt =</span> snpsToAdd<span class="op">$</span>alt
    )
  }
  
  <span class="co"># select indels and insert to foo ## Indels could be filtered here, too</span>
  toAdd &lt;-<span class="st"> </span><span class="kw">select</span>(indels, <span class="kw">c</span>(chr, pos, ref, alt)) <span class="op">%&gt;%</span>
<span class="st">    </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="kw">between</span>(pos, unit_defs[rowIndex, ]<span class="op">$</span>agg_start, unit_defs[rowIndex, ]<span class="op">$</span>agg_end)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># to vectorize</span>
<span class="st">    </span><span class="kw">distinct</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">group_id =</span> unit_defs[rowIndex, ]<span class="op">$</span>gene_id)
  
  <span class="cf">if</span> (rowIndex <span class="op">%%</span><span class="st"> </span><span class="dv">10</span> <span class="op">==</span><span class="st"> </span><span class="dv">0</span>){
    <span class="kw">message</span>(
      <span class="kw">paste0</span>(<span class="st">&quot;row: &quot;</span>, rowIndex, 
             <span class="st">&quot; snps to add: &quot;</span>, <span class="kw">nrow</span>(snpsToAdd), 
             <span class="st">&quot; indels to add: &quot;</span>, <span class="kw">nrow</span>(toAdd)))
  }
  
  <span class="cf">if</span> (<span class="kw">nrow</span>(toAdd) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) {
    foo &lt;-<span class="st"> </span><span class="kw">add_row</span>(
      foo,
      <span class="dt">group_id =</span> toAdd<span class="op">$</span>group_id,
      <span class="dt">chromosome =</span> toAdd<span class="op">$</span>chr,
      <span class="dt">position =</span> toAdd<span class="op">$</span>pos,
      <span class="dt">ref =</span> toAdd<span class="op">$</span>ref,
      <span class="dt">alt =</span> toAdd<span class="op">$</span>alt
    )
  }
}

aggregated_variants &lt;-<span class="st"> </span><span class="kw">distinct</span>(foo)</code></pre></div>
<p>That may not be fast or pretty, but we’ve now got a set of variants aggregated into genic units using the GENCODE gene model! This set can be saved and used with the analysis pipeline for association testing.</p>
<p>We can inspect the tibble with glimpse:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">glimpse</span>(aggregated_variants)</code></pre></div>
<p>We can do things like counting how many genic units we’re using:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">distinct</span>(<span class="kw">as.tibble</span>(aggregated_variants<span class="op">$</span>group_id))</code></pre></div>
<p>We can look at number of variants per aggregation unit:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts &lt;-<span class="st"> </span>aggregated_variants <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(group_id) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="kw">n</span>())</code></pre></div>
<p>Feel free to look at other summary statistics and do other exploratory data analysis as you’d like, but don’t forget to save it if you’d like to use it for the analysis pipeline!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(aggregated_variants, <span class="dt">file =</span> <span class="st">&quot;chr22_gene_aggregates.RDA&quot;</span>)</code></pre></div>
</div>
<div id="aggregate-unit-for-association-testing-exercise" class="section level3">
<h3><span class="header-section-number">7.5.4</span> Aggregate unit for association testing exercise</h3>
<p>We will be using a slightly different gene-based aggregation unit for the assocation testing exercise. As before, this analysis uses a subset of the TOPMed SNP variants that are present in the 1000 Genomes Project. However, in this exercise, the genic units include TOPMed SNP variants from all chromosomes (no indels, and not just chromosome 22 as before). Further, each genic unit is expanded to include the set of TOPMed SNP variants falling within a GENCODE-defined gene along with 20 kb flanking regions upstream and downstream of that range.</p>
<p>In a larger-scale analysis of TOPMed data, aggregation units could include both TOPMed SNP and indel variants falling within defined aggregation units, and would not be restricted to the variants also present in this chosen subset of the 1000 Genomes Project. An analyst might also choose to filter variants within each unit based on various annotations (examples include loss of function, conservation, deleteriousness scores, etc.).</p>
<p>As before, the aggregation units are defined in an R dataframe. Each row of the dataframe specifies a variant (chromosome, position, ref, alt) and the group identifier (group_id) assigned to it. Mutiple rows with different group identifiers can be specified to assign a variant to different groups (for example a variant can be assigned to mutiple genes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aggfile &lt;-<span class="st"> &quot;variants_by_gene.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(aggfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(workshop.path, aggfile), aggfile)
aggunit &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(aggfile)
<span class="kw">names</span>(aggunit)
<span class="kw">head</span>(aggunit)

<span class="co"># an example of variant that is present in mutiple groups</span>
<span class="kw">library</span>(dplyr)
mult &lt;-<span class="st"> </span>aggunit <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">group_by</span>(chr, pos) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">summarise</span>(<span class="dt">n=</span><span class="kw">n</span>()) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(n <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">inner_join</span>(aggunit, mult[<span class="dv">2</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</code></pre></div>
</div>
<div id="association-testing-with-aggregate-units" class="section level3">
<h3><span class="header-section-number">7.5.5</span> Association testing with aggregate units</h3>
<p>We can run a burden test or SKAT on each of these units using <code>assocTestAggregate</code>. We define a <code>SeqVarListIterator</code> object where each list element is an aggregate unit. The constructor expects a <code>GRangesList</code>, so we use the TopmedPipeline function <code>aggregateGRangesList</code> to quickly convert our single dataframe to the required format. This function can account for multiallelic variants (the same chromosome, position, and ref, but different alt alleles).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TopmedPipeline)
aggVarList &lt;-<span class="st"> </span><span class="kw">aggregateGRangesList</span>(aggunit)
<span class="kw">length</span>(aggVarList)
<span class="kw">head</span>(<span class="kw">names</span>(aggVarList))
aggVarList[[<span class="dv">1</span>]]

<span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarListIterator</span>(seqData, <span class="dt">variantRanges=</span>aggVarList, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>As in the previous section, we must fit the null model before running the association test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;Burden&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">names</span>(assoc)
<span class="kw">head</span>(assoc<span class="op">$</span>results)
<span class="kw">head</span>(<span class="kw">names</span>(assoc<span class="op">$</span>variantInfo))
<span class="kw">head</span>(assoc<span class="op">$</span>variantInfo[[<span class="dv">1</span>]])

<span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>Score.pval)</code></pre></div>
</div>
<div id="exercise-2" class="section level3">
<h3><span class="header-section-number">7.5.6</span> Exercise</h3>
<p>Since we are working with a subset of the data, many of the genes listed in <code>group_id</code> have a very small number of variants. Create a new set of units based on position rather than gene name, using the TopmedPipeline function <code>aggregateGRanges</code>. Then run SKAT using those units and a <code>SeqVarRangeIterator</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">agg2 &lt;-<span class="st"> </span>aggunit <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">chr=</span><span class="kw">factor</span>(chr, <span class="dt">levels=</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>, <span class="st">&quot;X&quot;</span>))) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">select</span>(chr, pos) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">distinct</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">group_by</span>(chr) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">summarise</span>(<span class="dt">min=</span><span class="kw">min</span>(pos), <span class="dt">max=</span><span class="kw">max</span>(pos))
<span class="kw">head</span>(agg2)

aggByPos &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(<span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(agg2), <span class="cf">function</span>(i) {
    <span class="kw">data.frame</span>(<span class="dt">chr=</span>agg2<span class="op">$</span>chr[i],
               <span class="dt">start=</span><span class="kw">seq</span>(agg2<span class="op">$</span>min[i], agg2<span class="op">$</span>max[i]<span class="op">-</span><span class="fl">1e6</span>, <span class="dt">length.out=</span><span class="dv">10</span>),
               <span class="dt">end=</span><span class="kw">seq</span>(agg2<span class="op">$</span>min[i]<span class="op">+</span><span class="fl">1e6</span>, agg2<span class="op">$</span>max[i], <span class="dt">length.out=</span><span class="dv">10</span>))
})) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">group_id=</span><span class="dv">1</span><span class="op">:</span><span class="kw">n</span>())
<span class="kw">head</span>(aggByPos)

aggVarList &lt;-<span class="st"> </span><span class="kw">aggregateGRanges</span>(aggByPos)
aggVarList[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]

<span class="kw">seqResetFilter</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarRangeIterator</span>(seqData, <span class="dt">variantRanges=</span>aggVarList, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestAggregate</span>(iterator, nullmod, <span class="dt">test=</span><span class="st">&quot;SKAT&quot;</span>, <span class="dt">AF.max=</span><span class="fl">0.1</span>, <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">25</span>))
<span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="finding-unharmonized-topmed-study-phenotypes-on-dbgap.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="analysis-pipeline.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
