<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>TOPMed Analysis Workshop</title>
  <meta name="description" content="Course materials for the TOPMed analysis workshop">
  <meta name="generator" content="bookdown 0.4 and GitBook 2.6.7">

  <meta property="og:title" content="TOPMed Analysis Workshop" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course materials for the TOPMed analysis workshop" />
  <meta name="github-repo" content="UW-GAC/topmed_workshop_2017" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="TOPMed Analysis Workshop" />
  
  <meta name="twitter:description" content="Course materials for the TOPMed analysis workshop" />
  



<meta name="date" content="2017-08-03">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="finding-topmed-study-phenotypes-on-dbgap.html">
<link rel="next" href="analysis-pipeline.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html"><i class="fa fa-check"></i><b>2</b> Downloading Data From The TOPMed Exchange Area</a><ul>
<li class="chapter" data-level="2.1" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#what-is-the-topmed-exchange-area-ea"><i class="fa fa-check"></i><b>2.1</b> What is the TOPMed Exchange Area (EA)?</a></li>
<li class="chapter" data-level="2.2" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#how-does-the-ea-work"><i class="fa fa-check"></i><b>2.2</b> How does the EA work?</a></li>
<li class="chapter" data-level="2.3" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#what-data-can-i-find-in-the-ea"><i class="fa fa-check"></i><b>2.3</b> What data can I find in the EA?</a></li>
<li class="chapter" data-level="2.4" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#who-can-access-the-topmed-ea"><i class="fa fa-check"></i><b>2.4</b> Who can access the TOPMed EA?</a></li>
<li class="chapter" data-level="2.5" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#how-can-i-download-data-from-the-ea"><i class="fa fa-check"></i><b>2.5</b> How can I download data from the EA?</a></li>
<li class="chapter" data-level="2.6" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#quick-how-to"><i class="fa fa-check"></i><b>2.6</b> Quick How-to</a></li>
<li class="chapter" data-level="2.7" data-path="downloading-data-from-the-topmed-exchange-area.html"><a href="downloading-data-from-the-topmed-exchange-area.html#useful-resources"><i class="fa fa-check"></i><b>2.7</b> Useful resources</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="gds-format.html"><a href="gds-format.html"><i class="fa fa-check"></i><b>3</b> GDS format</a><ul>
<li class="chapter" data-level="3.1" data-path="gds-format.html"><a href="gds-format.html#exploring-a-gds-file"><i class="fa fa-check"></i><b>3.1</b> Exploring a GDS file</a></li>
<li class="chapter" data-level="3.2" data-path="gds-format.html"><a href="gds-format.html#exercise-hwe"><i class="fa fa-check"></i><b>3.2</b> Exercise: HWE</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="computing-a-grm.html"><a href="computing-a-grm.html"><i class="fa fa-check"></i><b>4</b> Computing a GRM</a></li>
<li class="chapter" data-level="5" data-path="pc-relate.html"><a href="pc-relate.html"><i class="fa fa-check"></i><b>5</b> PC-Relate</a><ul>
<li class="chapter" data-level="5.1" data-path="pc-relate.html"><a href="pc-relate.html#king"><i class="fa fa-check"></i><b>5.1</b> KING</a></li>
<li class="chapter" data-level="5.2" data-path="pc-relate.html"><a href="pc-relate.html#pc-air"><i class="fa fa-check"></i><b>5.2</b> PC-AiR</a></li>
<li class="chapter" data-level="5.3" data-path="pc-relate.html"><a href="pc-relate.html#pc-relate-1"><i class="fa fa-check"></i><b>5.3</b> PC-Relate</a></li>
<li class="chapter" data-level="5.4" data-path="pc-relate.html"><a href="pc-relate.html#exercise"><i class="fa fa-check"></i><b>5.4</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html"><i class="fa fa-check"></i><b>6</b> Phenotype Harmonization</a><ul>
<li class="chapter" data-level="6.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#exercise-overview-run-diagnostics-on-a-harmonized-phenotype"><i class="fa fa-check"></i><b>6.1</b> Exercise overview: run diagnostics on a harmonized phenotype</a></li>
<li class="chapter" data-level="6.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#import-phentoype-files-into-r"><i class="fa fa-check"></i><b>6.2</b> Import phentoype files into R</a></li>
<li class="chapter" data-level="6.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#run-some-quick-diagnostics-on-the-files"><i class="fa fa-check"></i><b>6.3</b> Run some quick diagnostics on the files</a></li>
<li class="chapter" data-level="6.4" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-a-mixed-model-for-statistical-comparisons"><i class="fa fa-check"></i><b>6.4</b> Fit a mixed model for statistical comparisons</a><ul>
<li class="chapter" data-level="6.4.1" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#create-an-annotated-data-frame"><i class="fa fa-check"></i><b>6.4.1</b> Create an Annotated Data Frame</a></li>
<li class="chapter" data-level="6.4.2" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#fit-the-mixed-model"><i class="fa fa-check"></i><b>6.4.2</b> Fit the mixed model</a></li>
<li class="chapter" data-level="6.4.3" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#model-interpretation"><i class="fa fa-check"></i><b>6.4.3</b> Model interpretation</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="phenotype-harmonization.html"><a href="phenotype-harmonization.html#final-considerations"><i class="fa fa-check"></i><b>6.5</b> Final considerations</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="finding-topmed-study-phenotypes-on-dbgap.html"><a href="finding-topmed-study-phenotypes-on-dbgap.html"><i class="fa fa-check"></i><b>7</b> Finding TOPMed Study Phenotypes on dbGaP</a></li>
<li class="chapter" data-level="8" data-path="association-tests.html"><a href="association-tests.html"><i class="fa fa-check"></i><b>8</b> Association tests</a><ul>
<li class="chapter" data-level="8.1" data-path="association-tests.html"><a href="association-tests.html#null-model"><i class="fa fa-check"></i><b>8.1</b> Null model</a></li>
<li class="chapter" data-level="8.2" data-path="association-tests.html"><a href="association-tests.html#single-variant-tests"><i class="fa fa-check"></i><b>8.2</b> Single-variant tests</a></li>
<li class="chapter" data-level="8.3" data-path="association-tests.html"><a href="association-tests.html#sliding-window-tests"><i class="fa fa-check"></i><b>8.3</b> Sliding window tests</a></li>
<li class="chapter" data-level="8.4" data-path="association-tests.html"><a href="association-tests.html#exercise-logistic-regression"><i class="fa fa-check"></i><b>8.4</b> Exercise: logistic regression</a></li>
<li class="chapter" data-level="8.5" data-path="association-tests.html"><a href="association-tests.html#aggregate-tests"><i class="fa fa-check"></i><b>8.5</b> Aggregate tests</a><ul>
<li class="chapter" data-level="8.5.1" data-path="association-tests.html"><a href="association-tests.html#variant-annotation"><i class="fa fa-check"></i><b>8.5.1</b> Variant annotation</a></li>
<li class="chapter" data-level="8.5.2" data-path="association-tests.html"><a href="association-tests.html#defining-aggregate-units"><i class="fa fa-check"></i><b>8.5.2</b> Defining aggregate units</a></li>
<li class="chapter" data-level="8.5.3" data-path="association-tests.html"><a href="association-tests.html#association-testing-with-aggregate-units"><i class="fa fa-check"></i><b>8.5.3</b> Association testing with aggregate units</a></li>
<li class="chapter" data-level="8.5.4" data-path="association-tests.html"><a href="association-tests.html#exercise-1"><i class="fa fa-check"></i><b>8.5.4</b> Exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html"><i class="fa fa-check"></i><b>9</b> Analysis Pipeline</a><ul>
<li class="chapter" data-level="9.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-a-local-cluster"><i class="fa fa-check"></i><b>9.1</b> Running on a local cluster</a></li>
<li class="chapter" data-level="9.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#running-on-aws-batch"><i class="fa fa-check"></i><b>9.2</b> Running on AWS Batch</a><ul>
<li class="chapter" data-level="9.2.1" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#log-into-aws-docker-image"><i class="fa fa-check"></i><b>9.2.1</b> Log into AWS docker image</a></li>
<li class="chapter" data-level="9.2.2" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#cd-to-working-directory-and-create-config-file"><i class="fa fa-check"></i><b>9.2.2</b> cd to working directory and create config file</a></li>
<li class="chapter" data-level="9.2.3" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#print-out-aws-commands-if-executing-the-pipeline"><i class="fa fa-check"></i><b>9.2.3</b> Print out AWS commands if executing the pipeline</a></li>
<li class="chapter" data-level="9.2.4" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#execute-the-pipeline"><i class="fa fa-check"></i><b>9.2.4</b> Execute the pipeline</a></li>
<li class="chapter" data-level="9.2.5" data-path="analysis-pipeline.html"><a href="analysis-pipeline.html#monitor-the-jobs"><i class="fa fa-check"></i><b>9.2.5</b> Monitor the jobs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="analysis-commons.html"><a href="analysis-commons.html"><i class="fa fa-check"></i><b>10</b> Analysis Commons</a><ul>
<li class="chapter" data-level="10.1" data-path="analysis-commons.html"><a href="analysis-commons.html#outline"><i class="fa fa-check"></i><b>10.1</b> Outline</a></li>
<li class="chapter" data-level="10.2" data-path="analysis-commons.html"><a href="analysis-commons.html#web-interface-and-running-an-analysis-application"><i class="fa fa-check"></i><b>10.2</b> Web Interface and Running an Analysis Application</a><ul>
<li class="chapter" data-level="10.2.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-1-run-a-single-variant-analysis."><i class="fa fa-check"></i><b>10.2.1</b> Exercise 1) Run a single variant analysis.</a></li>
<li class="chapter" data-level="10.2.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-2-run-skat-test-grouping-variants-into-gene-transcript-regions-and-limit-the-variants-to-those-with-a-cadd-phred-score-2-and-maf-5."><i class="fa fa-check"></i><b>10.2.2</b> Exercise 2) Run SKAT test grouping variants into gene transcript regions and limit the variants to those with a CADD phred score &gt; 2 and MAF &lt;= 5%.</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="analysis-commons.html"><a href="analysis-commons.html#command-line-interface"><i class="fa fa-check"></i><b>10.3</b> Command line interface</a><ul>
<li class="chapter" data-level="10.3.1" data-path="analysis-commons.html"><a href="analysis-commons.html#log-in-to-awi"><i class="fa fa-check"></i><b>10.3.1</b> Log in to AWI</a></li>
<li class="chapter" data-level="10.3.2" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-3-navigate-directories-make-output-directory-examine-files"><i class="fa fa-check"></i><b>10.3.2</b> Exercise 3) Navigate directories, make output directory, examine files</a></li>
<li class="chapter" data-level="10.3.3" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-4-run-single-variant-analysis-from-command-line-using-bash-script"><i class="fa fa-check"></i><b>10.3.3</b> Exercise 4) Run single variant analysis from command line using bash script</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="analysis-commons.html"><a href="analysis-commons.html#writing-your-own-apps"><i class="fa fa-check"></i><b>10.4</b> Writing your own Apps</a><ul>
<li class="chapter" data-level="10.4.1" data-path="analysis-commons.html"><a href="analysis-commons.html#exercise-5-write-an-app-that-creates-phenotype-residuals-and-performs-an-inverse-normal-transform"><i class="fa fa-check"></i><b>10.4.1</b> Exercise 5) Write an App that creates phenotype residuals and performs an inverse normal transform</a></li>
<li class="chapter" data-level="10.4.2" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-6-make-qq-plot"><i class="fa fa-check"></i><b>10.4.2</b> Optional Exercise 6) Make QQ plot</a></li>
<li class="chapter" data-level="10.4.3" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-7-run-conditional-analysis"><i class="fa fa-check"></i><b>10.4.3</b> Optional Exercise 7) Run conditional analysis</a></li>
<li class="chapter" data-level="10.4.4" data-path="analysis-commons.html"><a href="analysis-commons.html#optional-exercise-8-create-a-regional-association-plot-using-ld-extracted-from-your-data-set"><i class="fa fa-check"></i><b>10.4.4</b> Optional Exercise 8) Create a regional association plot using LD extracted from your data set</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">TOPMed Analysis Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="association-tests" class="section level1">
<h1><span class="header-section-number">8</span> Association tests</h1>
<p>Since TOPMed has many studies with related participants, we focus on linear mixed models. Logistic mixed models are also possible using GENESIS, see the <a href="https://www.ncbi.nlm.nih.gov/pubmed/27018471">GMMAT paper</a>.</p>
<div id="null-model" class="section level2">
<h2><span class="header-section-number">8.1</span> Null model</h2>
<p>The first step in an association test is to fit the null model. We use the <code>AnnotatedDataFrame</code> with phenotypes, and a GRM. If the sample set involves multiple distinct groups with different variances for the phenotype, we recommend allowing the model to use heterogeneous variance among groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data.path &lt;-<span class="st"> &quot;https://github.com/smgogarten/analysis_pipeline/raw/devel/testdata&quot;</span>
sampfile &lt;-<span class="st"> &quot;1KG_phase3_subset_annot.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(sampfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(data.path, sampfile), sampfile)
annot &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(sampfile)

grmfile &lt;-<span class="st"> &quot;grm.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(grmfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(data.path, grmfile), grmfile)
grm &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(grmfile)
<span class="kw">rownames</span>(grm<span class="op">$</span>grm) &lt;-<span class="st"> </span><span class="kw">colnames</span>(grm<span class="op">$</span>grm) &lt;-<span class="st"> </span>grm<span class="op">$</span>sample.id

<span class="kw">library</span>(GENESIS)
nullmod &lt;-<span class="st"> </span><span class="kw">fitNullMM</span>(annot, <span class="dt">outcome=</span><span class="st">&quot;outcome&quot;</span>, <span class="dt">covars=</span><span class="kw">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;Population&quot;</span>), 
                     <span class="dt">covMatList=</span>grm<span class="op">$</span>grm, <span class="dt">group.var=</span><span class="st">&quot;Population&quot;</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>We also recommend taking an inverse normal transform of the residuals and refitting the model. This is done separately for each group, and the transformed residuals are rescaled. See the full procedure in the<br />
<a href="https://github.com/smgogarten/analysis_pipeline#association-testing">pipeline documenation</a>.</p>
</div>
<div id="single-variant-tests" class="section level2">
<h2><span class="header-section-number">8.2</span> Single-variant tests</h2>
<p>Single-variant tests are the same as in GWAS. We use the <code>assocTestMM</code> function in GENESIS. We have to create a <code>SeqVarData</code> object including both the GDS file and the sample annotation containing phenotypes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqVarTools)
gdsfile &lt;-<span class="st"> &quot;1KG_phase3_subset_chr1.gds&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(gdsfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(data.path, gdsfile), gdsfile)
gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)
seqData &lt;-<span class="st"> </span><span class="kw">SeqVarData</span>(gds, <span class="dt">sampleData=</span>annot)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestMM</span>(seqData, nullmod)
<span class="kw">head</span>(assoc)</code></pre></div>
<pre><code>##   snpID chr    n          MAF minor.allele          Est        SE
## 1     1   1 1126 0.0039964476          alt  0.037896432 0.3547545
## 2     2   1 1126 0.0492895204          alt  0.173271721 0.1038710
## 3     3   1 1126 0.0004440497          alt  0.034881381 1.0211652
## 4     4   1 1126 0.0008880995          alt  0.003698581 0.6811618
## 5     5   1 1126 0.0071047957          alt -0.062695115 0.2685319
## 6     6   1 1126 0.0022202487          alt  0.574104228 0.4504458
##      Wald.Stat  Wald.pval
## 1 1.141145e-02 0.91492830
## 2 2.782705e+00 0.09528713
## 3 1.166797e-03 0.97275083
## 4 2.948287e-05 0.99566766
## 5 5.450992e-02 0.81539366
## 6 1.624413e+00 0.20247756</code></pre>
<p>We make a QQ plot to examine the results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
qqPlot &lt;-<span class="st"> </span><span class="cf">function</span>(pval) {
    pval &lt;-<span class="st"> </span>pval[<span class="op">!</span><span class="kw">is.na</span>(pval)]
    n &lt;-<span class="st"> </span><span class="kw">length</span>(pval)
    x &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>n
    dat &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">obs=</span><span class="kw">sort</span>(pval),
                      <span class="dt">exp=</span>x<span class="op">/</span>n,
                      <span class="dt">upper=</span><span class="kw">qbeta</span>(<span class="fl">0.025</span>, x, <span class="kw">rev</span>(x)),
                      <span class="dt">lower=</span><span class="kw">qbeta</span>(<span class="fl">0.975</span>, x, <span class="kw">rev</span>(x)))
    
    <span class="kw">ggplot</span>(dat, <span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(obs))) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(upper)), <span class="dt">color=</span><span class="st">&quot;gray&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="op">-</span><span class="kw">log10</span>(exp), <span class="op">-</span><span class="kw">log10</span>(lower)), <span class="dt">color=</span><span class="st">&quot;gray&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">        </span><span class="kw">geom_abline</span>(<span class="dt">intercept=</span><span class="dv">0</span>, <span class="dt">slope=</span><span class="dv">1</span>, <span class="dt">color=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span>
<span class="st">        </span><span class="kw">xlab</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="op">-</span>log[<span class="dv">10</span>], <span class="st">&quot;(expected P)&quot;</span>))) <span class="op">+</span>
<span class="st">        </span><span class="kw">ylab</span>(<span class="kw">expression</span>(<span class="kw">paste</span>(<span class="op">-</span>log[<span class="dv">10</span>], <span class="st">&quot;(observed P)&quot;</span>))) <span class="op">+</span>
<span class="st">        </span><span class="kw">theme_bw</span>()
}    

<span class="kw">qqPlot</span>(assoc<span class="op">$</span>Wald.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_single_qq-1.png" width="672" /></p>
</div>
<div id="sliding-window-tests" class="section level2">
<h2><span class="header-section-number">8.3</span> Sliding window tests</h2>
<p>For rare variants, we can do burden tests or SKAT on sliding windows using the GENESIS function <code>assocTestSeqWindow</code>. We restrict the test to variants with alternate allele frequency &lt; 0.1. (For real data, this threshold would be lower.) We use a flat weighting scheme.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assoc &lt;-<span class="st"> </span><span class="kw">assocTestSeqWindow</span>(seqData, nullmod, <span class="dt">test=</span><span class="st">&quot;Burden&quot;</span>, <span class="dt">AF.range=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.1</span>),
                            <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">window.size=</span><span class="dv">5</span>, <span class="dt">window.shift=</span><span class="dv">2</span>)
<span class="kw">names</span>(assoc)</code></pre></div>
<pre><code>## [1] &quot;param&quot;       &quot;window&quot;      &quot;nsample&quot;     &quot;results&quot;     &quot;variantInfo&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr window.start window.stop n.site dup burden.skew       Score
## 1   1       966001      971000      1   0   11.036036  0.30138968
## 2   1       968001      973000      1   1   11.036036  0.30138968
## 3   1       970001      975000      1   1   11.036036  0.30138968
## 4   1       982001      987000      1   0    3.041979 16.03409196
## 5   1       984001      989000      1   1    3.041979 16.03409196
## 6   1      1022001     1027000      1   0   33.466573  0.03348047
##          Var  Score.stat Score.pval
## 1  7.9529830 0.011421594 0.91489064
## 2  7.9529830 0.011421594 0.91489064
## 3  7.9529830 0.011421594 0.91489064
## 4 92.5372698 2.778254702 0.09555224
## 5 92.5372698 2.778254702 0.09555224
## 6  0.9598379 0.001167845 0.97273860</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo)</code></pre></div>
<pre><code>##   variantID allele chr     pos n.obs         freq weight
## 1         1      1   1  970546  1126 0.0039964476      1
## 2         2      1   1  985900  1126 0.0492895204      1
## 3         3      1   1 1025045  1126 0.0004440497      1
## 4         4      1   1 1265550  1126 0.0008880995      1
## 5         5      1   1 1472676  1126 0.0071047957      1
## 6         6      1   1 1735725  1126 0.0022202487      1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>Score.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_window_burden-1.png" width="672" /></p>
<p>For SKAT, we use the Wu weights.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assoc &lt;-<span class="st"> </span><span class="kw">assocTestSeqWindow</span>(seqData, nullmod, <span class="dt">test=</span><span class="st">&quot;SKAT&quot;</span>, <span class="dt">AF.range=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.1</span>),
                            <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">25</span>), <span class="dt">window.size=</span><span class="dv">5</span>, <span class="dt">window.shift=</span><span class="dv">2</span>)
<span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##   chr window.start window.stop n.site dup          Q_0     pval_0 err_0
## 1   1       966001      971000      1   0 4.684458e+01 0.91489064     0
## 2   1       968001      973000      1   1 4.684458e+01 0.91489064     0
## 3   1       970001      975000      1   1 4.684458e+01 0.91489064     0
## 4   1       982001      987000      1   0 1.419993e+04 0.09555224     0
## 5   1       984001      989000      1   1 1.419993e+04 0.09555224     0
## 6   1      1022001     1027000      1   0 6.858109e-01 0.97273860     0</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo)</code></pre></div>
<pre><code>##   variantID allele chr     pos n.obs         freq    weight
## 1         1      1   1  970546  1126 0.0039964476 22.709172
## 2         2      1   1  985900  1126 0.0492895204  7.431881
## 3         3      1   1 1025045  1126 0.0004440497 24.734926
## 4         4      1   1 1265550  1126 0.0008880995 24.472547
## 5         5      1   1 1472676  1126 0.0071047957 21.067933
## 6         6      1   1 1735725  1126 0.0022202487 23.701317</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>pval_<span class="dv">0</span>)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_window_skat-1.png" width="672" /></p>
</div>
<div id="exercise-logistic-regression" class="section level2">
<h2><span class="header-section-number">8.4</span> Exercise: logistic regression</h2>
<p><code>fitNullMM</code> can use a binary phenotype as the outcome variable by specifying the argument <code>family=binomial</code>. Use the <code>status</code> column in the sample annotation to fit a null model for simulated case/control status. Then run a single-variant test and a sliding window test using this model.</p>
</div>
<div id="aggregate-tests" class="section level2">
<h2><span class="header-section-number">8.5</span> Aggregate tests</h2>
<div id="variant-annotation" class="section level3">
<h3><span class="header-section-number">8.5.1</span> Variant annotation</h3>
<p>Rare variants are generally aggregated into some meaningful units for association testing to decrease multiple testing burden and increase statistical power. Various genomic and epigenomic annotations can be used to define aggregation units and filter them. A large number of annotations are available through the Whole Genome Sequence Annotator (WGSA) to the TOPMed users.</p>
</div>
<div id="defining-aggregate-units" class="section level3">
<h3><span class="header-section-number">8.5.2</span> Defining aggregate units</h3>
<p>We will be using a gene-based aggregation unit, where each unit is a GENCODE gene and 20 kb flanking region upstream and downstream of it. For real data, one will likely filter variants within each unit based on various annotations (examples include loss of function, conservation, deleteriousness scores, etc.).</p>
<p>The aggregation units are defined in an R dataframe. Each row of the dataframe specifies a variant (chromosome, position, ref, alt) and the group identifier (group_id) assigned to it. Mutiple rows with different group identifiers can be specified to assign a variant to different groups (for example a variant can be assigned to mutiple genes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aggfile &lt;-<span class="st"> &quot;variants_by_gene.RData&quot;</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(aggfile)) <span class="kw">download.file</span>(<span class="kw">file.path</span>(workshop.path, aggfile), aggfile)
aggunit &lt;-<span class="st"> </span>TopmedPipeline<span class="op">::</span><span class="kw">getobj</span>(aggfile)
<span class="kw">names</span>(aggunit)</code></pre></div>
<pre><code>## [1] &quot;group_id&quot;   &quot;chromosome&quot; &quot;position&quot;   &quot;ref&quot;        &quot;alt&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(aggunit)</code></pre></div>
<pre><code>## # A tibble: 6 x 5
##             group_id chromosome position   ref   alt
##                &lt;chr&gt;      &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;
## 1  ENSG00000000005.5          X 99850725     A     G
## 2  ENSG00000000938.8          1 27960254     A     G
## 3  ENSG00000001084.6          6 53357691     C     T
## 4  ENSG00000001084.6          6 53413986     T     C
## 5  ENSG00000001084.6          6 53466979     C     T
## 6 ENSG00000001167.10          6 41064020     A     G</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># an example of variant that is present in mutiple groups</span>
<span class="kw">library</span>(dplyr)
mult &lt;-<span class="st"> </span>aggunit <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">group_by</span>(chromosome, position) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">summarise</span>(<span class="dt">n=</span><span class="kw">n</span>()) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">filter</span>(n <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw">inner_join</span>(aggunit, mult[<span class="dv">2</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>])</code></pre></div>
<pre><code>## # A tibble: 4 x 5
##            group_id chromosome position   ref   alt
##               &lt;chr&gt;      &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; &lt;chr&gt;
## 1 ENSG00000188157.9          1   985900     C     T
## 2 ENSG00000217801.5          1   985900     C     T
## 3 ENSG00000242590.1          1   985900     C     T
## 4 ENSG00000273443.1          1   985900     C     T</code></pre>
</div>
<div id="association-testing-with-aggregate-units" class="section level3">
<h3><span class="header-section-number">8.5.3</span> Association testing with aggregate units</h3>
<p>We can run a burden test or SKAT on each of these units using the GENESIS function <code>assocTestSeq</code>. This function expects a list, where each element of the list is a dataframe representing a single aggregation unit and containing the unique variant.id assigned to each variant in a GDS file. We use the TopmedPipeline function <code>aggregateListByAllele</code> to quickly convert our single dataframe to the required format. This function can account for multiallelic variants (the same chromosome, position, and ref, but different alt alleles). The first argument is the GDS object returned by <code>seqOpen</code> (see above).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TopmedPipeline)
aggVarList &lt;-<span class="st"> </span><span class="kw">aggregateListByAllele</span>(gds, aggunit)
<span class="kw">length</span>(aggVarList)</code></pre></div>
<pre><code>## [1] 932</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">names</span>(aggVarList))</code></pre></div>
<pre><code>## [1] &quot;ENSG00000188157.9&quot;  &quot;ENSG00000242590.1&quot;  &quot;ENSG00000217801.5&quot; 
## [4] &quot;ENSG00000273443.1&quot;  &quot;ENSG00000131591.13&quot; &quot;ENSG00000237330.2&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aggVarList[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>##   variant.id chromosome position ref nAlleles allele allele.index
## 1          1          1   970546   C        2      G            1
## 2          2          1   985900   C        2      T            1</code></pre>
<p>As in the previous section, we must fit the null model before running the association test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assoc &lt;-<span class="st"> </span><span class="kw">assocTestSeq</span>(seqData, nullmod, <span class="dt">test=</span><span class="st">&quot;Burden&quot;</span>, <span class="dt">aggVarList=</span>aggVarList, 
                      <span class="dt">AF.range=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="fl">0.1</span>), <span class="dt">weight.beta=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">names</span>(assoc)</code></pre></div>
<pre><code>## [1] &quot;param&quot;       &quot;nsample&quot;     &quot;results&quot;     &quot;variantInfo&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>results)</code></pre></div>
<pre><code>##                    n.site n.sample.alt burden.skew       Score        Var
## ENSG00000188157.9       2          116    2.867095 16.33548164 99.2100284
## ENSG00000242590.1       2          116    2.867095 16.33548164 99.2100284
## ENSG00000217801.5       1          107    3.041979 16.03409196 92.5372698
## ENSG00000273443.1       1          107    3.041979 16.03409196 92.5372698
## ENSG00000131591.13      1            1   33.466573  0.03348047  0.9598379
## ENSG00000237330.2       1            1   33.466573  0.03348047  0.9598379
##                     Score.stat Score.pval
## ENSG00000188157.9  2.689727688 0.10099707
## ENSG00000242590.1  2.689727688 0.10099707
## ENSG00000217801.5  2.778254702 0.09555224
## ENSG00000273443.1  2.778254702 0.09555224
## ENSG00000131591.13 0.001167845 0.97273860
## ENSG00000237330.2  0.001167845 0.97273860</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">names</span>(assoc<span class="op">$</span>variantInfo))</code></pre></div>
<pre><code>## [1] &quot;ENSG00000188157.9&quot;  &quot;ENSG00000242590.1&quot;  &quot;ENSG00000217801.5&quot; 
## [4] &quot;ENSG00000273443.1&quot;  &quot;ENSG00000131591.13&quot; &quot;ENSG00000237330.2&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(assoc<span class="op">$</span>variantInfo[[<span class="dv">1</span>]])</code></pre></div>
<pre><code>##   variantID allele chr    pos n.obs        freq weight
## 1         1      1   1 970546  1126 0.003996448      1
## 2         2      1   1 985900  1126 0.049289520      1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qqPlot</span>(assoc<span class="op">$</span>results<span class="op">$</span>Score.pval)</code></pre></div>
<p><img src="_main_files/figure-html/assoc_aggregate-1.png" width="672" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>
</div>
<div id="exercise-1" class="section level3">
<h3><span class="header-section-number">8.5.4</span> Exercise</h3>
<p>Since we are working with a subset of the data, many of the genes listed in <code>group_id</code> have a very small number of variants. Create a new set of units combining adjacent genes, and run SKAT using those units.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="finding-topmed-study-phenotypes-on-dbgap.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="analysis-pipeline.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
